### PR TITLE
User authentication

### PR DESCRIPTION
**Pull request by BrookJeynes**
## Notes
- Response messages are now optional
- Added route to get reviews by unit
- Added route for admins to create accounts
- Code cleanup

### CODE REVIEW

**Review by Tarang74 — APPROVED** [review]
Nicely done!

File: api/application/src/review/delete.rs [hunk 1]
```diff
@@ -1,30 +1,57 @@
...
+
+                return Err((Status::NotFound, Json(response)));
+            }
+            _ => {
+                panic!("Database error - {}", err);
+            }
+        },
+    };
+
+    if token.claims.sub != review.user_id && token.claims.role != Role::Admin {
```
Comment (rjboas) [inline api/application/src/review/delete.rs:33]: Check for admin before accessing DB.

File: api/application/src/review/update.rs [hunk 2]
```diff
@@ -43,12 +55,40 @@ pub fn update_review(
...
+
+                return Err((Status::NotFound, Json(response)));
+            }
+            _ => {
+                panic!("Database error - {}", err);
+            }
+        },
+    };
+
+    if token.claims.sub != db_review.user_id && token.claims.role != Role::Admin {
```
Comment (rjboas) [inline api/application/src/review/update.rs:85]: Check for admin before accessing DB.

File: api/application/src/user/read.rs [hunk 3]
```diff
@@ -1,18 +1,27 @@
...
     user_id: Uuid,
     state: &State<ServerState>,
-) -> Result<Json<GetUser>, NotFound<Json<ResponseMessage>>> {
+    token: JWT,
+) -> Result<Json<GetUser>, (Status, Json<ResponseMessage>)> {
     use domain::schema::users;
 
     let pooled = &mut state.db_pool.get().unwrap();
 
+    if token.claims.sub != user_id && token.claims.role != Role::Admin {
```
Comment (rjboas) [inline api/application/src/user/read.rs:18]: Check for admin before accessing DB.

File: api/application/src/user/update.rs [hunk 4]
```diff
@@ -1,21 +1,33 @@
...
-use rocket::response::status::{Created, NotFound};
+use rocket::http::Status;
+use rocket::response::status::Created;
 use rocket::serde::json::Json;
 use rocket::State;
 use shared::response_models::ResponseMessage;
+use shared::token::JWT;
 use uuid::Uuid;
 
 pub fn update_user(
```
Comment (rjboas) [inline api/application/src/user/update.rs:15]: User is currently able to update their own role from User to Admin as a check for this is missing.

File: api/application/src/review/create.rs [hunk 5]
```diff
@@ -1,22 +1,20 @@
...
+    token: JWT,
+) -> Result<Created<String>, (Status, Json<ResponseMessage>)> {
     use domain::schema::reviews;
 
     let pooled = &mut state.db_pool.get().unwrap();
 
     let review = review.into_inner();
+    let review = NewReviewWithId::new(token.claims.sub, review);
 
     let review_count: i64 = match pooled.transaction(|c| {
```
Comment (rjboas) [inline api/application/src/review/create.rs:52]: Review on a non-existent unit code leads to panic (due to foreign key constraint). Bug not introduced in this PR but just so we are aware.

**Review by rjboas — CHANGES_REQUESTED** [review]
At minimum need to fix privilege escalation before merge, otherwise I'm happy with how it's going.

File: api/application/src/review/delete.rs (hunk 1 above)
Comment (BrookJeynes) [reply api/application/src/review/delete.rs:33]: This happens as we need the user id tied to the review before we can verify who they are

File: api/application/src/review/update.rs (hunk 2 above)
Comment (BrookJeynes) [reply api/application/src/review/update.rs:85]: This happens as we need the user id tied to the review before we can verify who they are

File: api/application/src/review/create.rs [hunk 6]
```diff
@@ -40,6 +41,36 @@ pub fn create_review(
...
+
+                return Err((Status::NotFound, Json(response)));
+            }
+            _ => {
+                panic!("Database error - {}", err);
+            }
+        },
+    };
+
+    if unit_count > 0 {
```
Comment (rjboas) [inline api/application/src/review/create.rs]: "If unit_count is greater than zero then it doesn't exist" what?

File: api/application/src/review/create.rs [hunk 7]
```diff
@@ -40,6 +41,36 @@ pub fn create_review(
         return Err((Status::Conflict, Json(response)));
     }
 
+    let unit_count: i64 = match pooled.transaction(|c| {
```
Comment (rjboas) [inline api/application/src/review/create.rs]: This would ideally be done as the first thing to happen in this function.

**Review by rjboas — APPROVED** [review]
lgtm

**Review by JMTC-dev — APPROVED** [review]
The code achieves what it sets out to do, neatly done and great work.

**Pull request merged**